File: Source/Clients and OTR/MockTransportSession+OTR.m
// 
// Wire
// Copyright (C) 2016 Wire Swiss GmbH
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
// 


#import "MockTransportSession+OTR.h"
#import <WireMockTransport/WireMockTransport-Swift.h>


@implementation MockTransportSession (OTR)

- (MockUserClient *)otrMessageSender:(NSDictionary *)payload
{
    NSString *senderClientId = payload[@"sender"];
    NSFetchRequest *request = [NSFetchRequest fetchRequestWithEntityName:@"UserClient"];
    request.predicate = [NSPredicate predicateWithFormat:@"identifier == %@", senderClientId];
    MockUserClient *client = [self.managedObjectContext executeFetchRequestOrAssert:request].firstObject;
    return client;
}

- (NSDictionary *)missedClients:(NSDictionary *)recipients conversation:(MockConversation *)conversation sender:(MockUserClient *)sender onlyForUserId:(NSString *)onlyForUserId
{
    return [self missedClients:recipients users:conversation.activeUsers.set sender:sender onlyForUserId:onlyForUserId];
}

- (NSDictionary *)missedClients:(NSDictionary *)recipients sender:(MockUserClient *)sender onlyForUserId:(NSString *)onlyForUserId
{
    return [self missedClients:recipients users:self.selfUser.connectionsAndTeamMembers sender:sender onlyForUserId:onlyForUserId];
}

- (NSDictionary *)missedClients:(NSDictionary *)recipients users:(NSSet<MockUser *> *)users sender:(MockUserClient *)sender onlyForUserId:(NSString *)onlyForUserId
{
    NSMutableDictionary *missedClients = [NSMutableDictionary new];
    for (MockUser *user in users) {
        if (onlyForUserId != nil && ![[NSUUID uuidWithTransportString:user.identifier] isEqual:[NSUUID uuidWithTransportString:onlyForUserId]]) {
            continue;
        }
        NSArray *recipientClients = [recipients[user.identifier] allKeys];
        NSSet *userClients = [user.clients mapWithBlock:^id(MockUserClient *client) {
            if (client != sender) {
                return client.identifier;
            }
            return nil;
        }];
        
        NSMutableSet *userMissedClients = [userClients mutableCopy];
        [userMissedClients minusSet:[NSSet setWithArray:recipientClients]];
        if (userMissedClients.count > 0) {
            missedClients[user.identifier] = userMissedClients.allObjects;
        }
    }
    return missedClients;
}

- (NSDictionary *)redundantClients:(NSDictionary *)recipients conversation:(MockConversation *)conversation
{
    return [self redundantClients:recipients users:conversation.activeUsers.set];
}

- (NSDictionary *)redundantClients:(NSDictionary *)recipients
{
    return [self redundantClients:recipients users:self.selfUser.connectionsAndTeamMembers];
}

- (NSDictionary *)redundantClients:(NSDictionary *)recipients users:(NSSet<MockUser *> *)users
{
    NSMutableDictionary *redundantClients = [NSMutableDictionary new];
    for (NSString *userId in recipients) {
        NSDictionary *recipientPayload = recipients[userId];
        MockUser *user = [users filteredSetUsingPredicate:[NSPredicate predicateWithFormat:@"identifier == %@", userId]].anyObject;
        NSArray *recipientClients = [recipientPayload allKeys];
        NSSet *userClients = [user.clients mapWithBlock:^id(MockUserClient *client) {
            return client.identifier;
        }];
        NSMutableSet *redundantUserClients = [NSMutableSet setWithArray:recipientClients];
        [redundantUserClients minusSet:userClients];
        if (redundantUserClients.count > 0) {
            redundantClients[userId] = redundantUserClients.allObjects;
        }
    }
    return redundantClients;
}

- (MockUserClient *)otrMessageSenderFromClientId:(ZMClientId *)sender
{
    NSString *senderClientId = [NSString stringWithFormat:@"%llx", [@(sender.client) longLongValue]];
    NSFetchRequest *request = [NSFetchRequest fetchRequestWithEntityName:@"UserClient"];
    request.predicate = [NSPredicate predicateWithFormat:@"identifier == %@", senderClientId];
    MockUserClient *client = [self.managedObjectContext executeFetchRequestOrAssert:request].firstObject;
    return client;
}

- (NSDictionary *)missedClientsFromRecipients:(NSArray *)recipients conversation:(MockConversation *)conversation sender:(MockUserClient *)sender onlyForUserId:(NSString *)onlyForUserId
{
    return [self missedClientsFromRecipients:recipients users:conversation.activeUsers.set sender:sender onlyForUserId:onlyForUserId];
}

- (NSDictionary *)missedClientsFromRecipients:(NSArray *)recipients sender:(MockUserClient *)sender onlyForUserId:(NSString *)onlyForUserId
{
    return [self missedClientsFromRecipients:recipients users:self.selfUser.connectionsAndTeamMembers sender:sender onlyForUserId:onlyForUserId];
}

- (NSDictionary *)missedClientsFromRecipients:(NSArray *)recipients users:(NSSet<MockUser *> *)users sender:(MockUserClient *)sender onlyForUserId:(NSString *)onlyForUserId
{
    NSMutableDictionary *missedClients = [NSMutableDictionary new];

    for (MockUser *user in users) {
        if (onlyForUserId != nil && ![[NSUUID uuidWithTransportString:user.identifier] isEqual:[NSUUID uuidWithTransportString:onlyForUserId]]) {
            continue;
        }
        ZMUserEntry *userEntry = [[recipients filteredArrayUsingPredicate:[NSPredicate predicateWithBlock:^BOOL(ZMUserEntry  * _Nonnull evaluatedObject, NSDictionary<NSString *,id> * _Nullable __unused bindings) {
            NSUUID *uuid = [[NSUUID alloc] initWithUUIDBytes:evaluatedObject.user.uuid.bytes];
            NSUUID *userId = [NSUUID uuidWithTransportString:user.identifier];
            return [uuid isEqual:userId] && (onlyForUserId == nil || [[NSUUID uuidWithTransportString:onlyForUserId] isEqual:userId]);
        }]] firstObject];
        
        NSArray *recipientClients = [userEntry.clients mapWithBlock:^id(ZMClientEntry *clientEntry) {
            return [NSString stringWithFormat:@"%llx", [@(clientEntry.client.client) unsignedLongLongValue]];
        }];
        
        NSSet *userClients = [user.clients mapWithBlock:^id(MockUserClient *client) {
            if (client != sender) {
                return client.identifier;
            }
            return nil;
        }];
        
        NSMutableSet *userMissedClients = [userClients mutableCopy];
        [userMissedClients minusSet:[NSSet setWithArray:recipientClients]];
        if (userMissedClients.count > 0) {
            missedClients[user.identifier] = userMissedClients.allObjects;
        }
    }
    return missedClients;
}

- (NSDictionary *)redundantClientsFromRecipients:(NSArray *)recipients conversation:(MockConversation *)conversation
{
    return [self redundantClientsFromRecipients:recipients users:conversation.activeUsers.set];
}

- (NSDictionary *)redundantClientsFromRecipients:(NSArray *)recipients
{
    return [self redundantClientsFromRecipients:recipients users:self.selfUser.connectionsAndTeamMembers];
}

- (NSDictionary *)redundantClientsFromRecipients:(NSArray *)recipients users:(NSSet<MockUser *> *)users
{
    NSMutableDictionary *redundantClients = [NSMutableDictionary new];
    
    for (MockUser *user in users) {
        ZMUserEntry *userEntry = [[recipients filteredArrayUsingPredicate:[NSPredicate predicateWithBlock:^BOOL(ZMUserEntry  * _Nonnull evaluatedObject, NSDictionary<NSString *,id> * _Nullable __unused bindings) {
            NSUUID *uuid = [[NSUUID alloc] initWithUUIDBytes:evaluatedObject.user.uuid.bytes];
            NSUUID *userId = [NSUUID uuidWithTransportString:user.identifier];
            return [uuid isEqual:userId];
        }]] firstObject];
        
        NSArray *recipientClients = [userEntry.clients mapWithBlock:^id(ZMClientEntry *clientEntry) {
            return [NSString stringWithFormat:@"%llx", [@(clientEntry.client.client) unsignedLongLongValue]];
        }];
        
        NSSet *userClients = [user.clients mapWithBlock:^id(MockUserClient *client) {
            return client.identifier;
        }];
        
        NSMutableSet *redundantUserClients = [NSMutableSet setWithArray:recipientClients];
        [redundantUserClients minusSet:userClients];
        if (redundantUserClients.count > 0) {
            redundantClients[user.identifier] = redundantUserClients.allObjects;
        }
    }
    return redundantClients;
}

- (void)insertOTRMessageEventsToConversation:(MockConversation *)conversation
                              requestPayload:(NSDictionary *)requestPayload
                            createEventBlock:(MockEvent *(^)(MockUserClient *recipient, NSData *messageData))createEventBlock
{
    NSDictionary *recipients = requestPayload[@"recipients"];
    
    NSArray *activeClients = [conversation.activeUsers.array flattenWithBlock:^NSArray *(MockUser *user) {
        return user.clients.allObjects;
    }];
    
    for (MockUserClient *activeClient in activeClients) {
        NSString *keyPath = [NSString stringWithFormat:@"%@.%@", activeClient.user.identifier, activeClient.identifier];
        NSString *messageContent = [recipients valueForKeyPath:keyPath];
        if (messageContent == nil) {
            continue;
        }
        NSData *messageData = [[NSData alloc] initWithBase64EncodedString:messageContent options:0];
        createEventBlock(activeClient, messageData);
    }
}

- (void)insertOTRMessageEventsToConversation:(MockConversation *)conversation
                           requestRecipients:(NSArray *)recipients
                                senderClient:(MockUserClient *)senderClient
                            createEventBlock:(MockEvent *(^)(MockUserClient *recipient, NSData *messageData, NSData *decryptedData))createEventBlock;
{
    NSArray *activeClients = [conversation.activeUsers.array flattenWithBlock:^NSArray *(MockUser *user) {
        return [user.clients.allObjects mapWithBlock:^id(MockUserClient *client) {
            return client.identifier;
        }];
    }];
    
    NSFetchRequest *allClientsRequest = [NSFetchRequest fetchRequestWithEntityName:@"UserClient"];
    allClientsRequest.predicate = [NSPredicate predicateWithFormat:@"identifier IN %@", activeClients];
    NSArray *allClients = [self.managedObjectContext executeFetchRequestOrAssert:allClientsRequest];
    
    NSArray *clientsEntries = [recipients flattenWithBlock:^NSArray *(ZMUserEntry *userEntry) {
        return userEntry.clients;
    }];
    
    for (ZMClientEntry *clientEntry in clientsEntries) {
        MockUserClient *client = [[allClients filteredArrayUsingPredicate:[NSPredicate predicateWithBlock:^BOOL(MockUserClient * _Nonnull aClient, NSDictionary<NSString *,id> * _Nullable __unused bindings) {
            NSString *clientId = [NSString stringWithFormat:@"%llx", [@(clientEntry.client.client) unsignedLongLongValue]];
            return [aClient.identifier isEqual:clientId];
        }]] firstObject];

        if (client != nil) {
            createEventBlock(client, clientEntry.text, [MockUserClient decryptMessageWithData:clientEntry.text from:senderClient to:client]);
        }
    }
}

@end


@implementation ZMNewOtrMessage (OtrMessage)
@end

@implementation ZMOtrAssetMeta (OtrMessage)
@end

File: Source/Clients and OTR/MockTransportSession+OTR.swift
//
// Wire
// Copyright (C) 2020 Wire Swiss GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
//

import WireProtos

extension MockTransportSession {
    
    func otrMessageSender(fromClientId sender: ClientId) -> MockUserClient? {
        let senderClientId = String(format: "%llx", CLongLong(sender.client))
        
        let request = NSFetchRequest<NSFetchRequestResult>(entityName: "UserClient")
        request.predicate = NSPredicate(format: "identifier == %@", senderClientId)
        let client = managedObjectContext.executeFetchRequestOrAssert(request)?.first as? MockUserClient
        return client
    }
    
    func missedClients(fromRecipients recipients: [UserEntry], conversation: MockConversation, sender: MockUserClient, onlyForUserId: String?) -> [String: [String]] {
        let users = conversation.activeUsers.set as! Set<MockUser>
        return missedClients(fromRecipients: recipients, users: users, sender: sender, onlyForUserId: onlyForUserId)
    }
    
    func missedClients(fromRecipients recipients: [UserEntry], sender: MockUserClient, onlyForUserId: String?) -> [String: [String]] {
        return missedClients(fromRecipients: recipients, users: selfUser.connectionsAndTeamMembers, sender: sender, onlyForUserId: onlyForUserId)
    }
    
    /// Returns a list of missing clients in the conversation that were not included in the list of intendend recipients
    /// - Parameters:
    ///   - recipients: list of intended recipients
    ///   - users: list of users to get clients from
    ///   - sender: will be excluded from list
    ///   - onlyForUserId: if not nil, only return missing recipients matching this user ID
    /// - Returns: missing clients
    func missedClients(fromRecipients recipients: [UserEntry], users: Set<MockUser>, sender: MockUserClient, onlyForUserId: String?) -> [String: [String]] {
        var missedClients = [String: [String]]()
    
        for user in users {
            if let id = onlyForUserId, UUID(uuidString: user.identifier) != UUID(uuidString: id) {
                continue
            }
            
            let userEntry = recipients.first { recipient in
                guard
                    let uuid = UUID(data: recipient.user.uuid),
                    let userId = UUID(uuidString: user.identifier) else {
                        return false
                }
                return uuid == userId && (onlyForUserId == nil || UUID(uuidString: onlyForUserId!) == userId)
            }
            
            let recipientClients = userEntry?.clients.map { entry in
                return String(format: "%llx", CUnsignedLongLong(entry.client.client))
            }
            
            var userClients: Set<String> = Set(user.userClients.compactMap { client in
                if client != sender {
                    return client.identifier
                }
                return nil
            })
            
            if let recipientClients = recipientClients {
                userClients.subtract(recipientClients)
            }
            
            if !userClients.isEmpty {
                missedClients[user.identifier] = Array(userClients)
            }
        }
        return missedClients
    }
    
    func deletedClients(fromRecipients recipients: [UserEntry], conversation: MockConversation) -> [String: [String]] {
        let users = conversation.activeUsers.set as! Set<MockUser>
        return deletedClients(fromRecipients: recipients, users: users)
    }
    
    func deletedClients(fromRecipients recipients: [UserEntry]) -> [String: [String]] {
        return deletedClients(fromRecipients: recipients, users: selfUser.connectionsAndTeamMembers)
    }
    
    /// Returns a list of deleted clients for broascasting that were included in the list of intendend recipients
    /// - Parameters:
    ///   - recipients: list of intended recipients
    ///   - users: set of users to get clients from
    /// - Returns: deleted clients
    func deletedClients(fromRecipients recipients: [UserEntry], users: Set<MockUser>) -> [String: [String]] {
        var deletedClients = [String: [String]]()
        
        for user in users {
            guard let userEntry = recipients.first(where: { recipient in
                guard
                    let uuid = UUID(data: recipient.user.uuid),
                    let userId = UUID(uuidString: user.identifier) else {
                        return false
                }
                return uuid == userId
            }) else {
                continue
            }
            
            let recipientClients = userEntry.clients.map { entry in
                return String(format: "%llx", CUnsignedLongLong(entry.client.client))
            }
            
            let userClients: Set<String> = Set(user.userClients.compactMap { client in
                return client.identifier
            })
            
            var deletedUserClients = Set(recipientClients)
            deletedUserClients.subtract(userClients)
            
            if !deletedUserClients.isEmpty {
                deletedClients[user.identifier] = Array(deletedUserClients)
            }
        }
        
        return deletedClients
    }
    
    func insertOTRMessageEvents(toConversation conversation: MockConversation,
                                recipients: [UserEntry],
                                senderClient: MockUserClient,
                                createEventBlock: (MockUserClient, Data, Data) -> MockEvent) {
        
        let activeUsers = conversation.activeUsers.array as? [MockUser]
        guard let activeClients = activeUsers?.flatMap({ user in
            return user.userClients.compactMap { $0.identifier }
        }) else {
            return
        }
       
        let allClientsRequest = NSFetchRequest<NSFetchRequestResult>(entityName: "UserClient")
        allClientsRequest.predicate = NSPredicate(format: "identifier IN %@", activeClients)
        let allClients1 = managedObjectContext.executeFetchRequestOrAssert(allClientsRequest)
        let allClients = allClients1 as? [MockUserClient]
        
        let clientsEntries = recipients.flatMap { return $0.clients }
        
        for entry in clientsEntries {
            guard let client = allClients?.first(where: { client in
                let clientId = String(format: "%llx", CUnsignedLongLong(entry.client.client))
                return client.identifier == clientId
            }) else {
                return
            }
            
            let decryptedData = MockUserClient.decryptMessage(data: entry.text, from: senderClient, to: client)
            _ = createEventBlock(client, entry.text, decryptedData)
        }
    }
}
